cpp/src/arrow/array.cc

```diff
@@ -42,4 +37,24 @@ void Array::Init(const TypePtr& type, int32_t length, int32_t null_count,
   }
 }
 
+bool Array::Equals(const Array& other) const {
+  if (this == &other) return true;
+  if (null_count_ != other.null_count_) {
+    return false;
+  }
+  if (null_count_ > 0) {
+    return nulls_->Equals(*other.nulls_, util::bytes_for_bits(length_));
+  } else {
+    return true;
+  }
```

Comment:
is there any reason you aren't comparing length here as well?  I suppose its subsumed by the nulls_ check when there are nulls, but what happens if there aren't?